Starting point

abdullahgarcia opened the pull request:
- Supports internal instance, master, and node starting points.
- Note that the supported starting points with this change still require work: handle kubectl-access for internal instance and handle as-root for master/node.
- The previously mentioned functionality and the support for pod starting point will come in the next pull requests.

jondkent commented:
Hmm, errors:

`In attack/scripts/bashrc line 82:
        NODE_NUMBER="$(echo $NODE_TYPE | sed "s/node-id: 'master-//" | sed "s/'//")"
                            ^-- SC2086: Double quote to prevent globbing and word splitting.


In attack/scripts/bashrc line 84:
        ssh_master $NODE_NUMBER
                   ^-- SC2086: Double quote to prevent globbing and word splitting.


In attack/scripts/bashrc line 86:
        NODE_NUMBER="$(echo $NODE_TYPE | sed "s/node-id: 'node-//" | sed "s/'//")"
                            ^-- SC2086: Double quote to prevent globbing and word splitting.


In attack/scripts/bashrc line 88:
        ssh_node $NODE_NUMBER
                 ^-- SC2086: Double quote to prevent globbing and word splitting.

The command '/bin/sh -c hadolint Dockerfile                             &&  hadolint attack/Dockerfile                  && shellcheck scripts/*                         && shellcheck attack/scripts/*                  && shellcheck simulation-scripts/perturb.sh     && shellcheck kubesim' returned a non-zero code: 1
make: *** [docker-build] Error 1`

raoulmillais commented:
we should add this script to the github action that does shellcheck to get faster automatic feedback on pull requests

raoulmillais commented on attack/scripts/bashrc:
```diff
@@ -61,6 +61,44 @@ info() {
+
+starting_point() {
+  # Identify the starting point mode.
+  MODE="$(sed '3,3!d' /tasks.yaml)"
```
This would be more readable and more robust if we installed `yq` in the attack container and used that over sed

abdullahgarcia replied:
Please, see last comment.

raoulmillais commented on attack/scripts/bashrc:
```diff
@@ -61,6 +61,44 @@ info() {
+    ssh_internal
+  else
+    if [[ $MODE =~ "node" ]]; then
+      NODE_TYPE="$(sed '4,4!d' /tasks.yaml)"
```
As above - prefer `yq` over sed ... I think having `yq` installed in the attack container may prove useful for other reasons in the future too

abdullahgarcia replied:
Please, see last comment.

raoulmillais commented on attack/scripts/bashrc:
```diff
@@ -61,6 +61,44 @@ info() {
+  #Determine the type.
+  if [[ $MODE =~ "internal-instance" ]]; then
+    ssh_internal
+  else
```
If you `return` from the `if` block so that you don't have to use else and indent further which makes the code more readable.  See https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88 for a fuller rationale :)

abdullahgarcia replied:
Covered in the new commit.

raoulmillais commented on attack/scripts/bashrc:
```diff
@@ -61,6 +61,44 @@ info() {
+      if [[ $MODE =~ "pod" ]]; then
+        echo "Starting point: POD - PENDING IMPLEMENTATION"
+      else
+        echo "There's no starting point or the scenario configuration file is not following the expected format."
```
we should disambiguate these 2 conditions (no starting point vs bad format)  .. using yq could help here

abdullahgarcia replied:
Covered in the new commit.

raoulmillais commented:
All the places where you're using sed yq qould be more readbable - particularly where you're doing string parsing and piping sed into sed :)

abdullahgarcia commented:
> All the places where you're using `sed`; yq would be more readable - particularly where you're doing string parsing and piping `sed` into `sed` :)

Fair points Raoul.

I'll correct this in the second part of the change if that's ok.

raoulmillais commented:
As discussed in private chat, let's build on this pull request and merge once the feature is complete.

abdullahgarcia commented:
As highlighted in the commit message:

- Added code to ensure lower case when retrieving the mode.
- Changed logic handling as per Raoul's suggestions.
- Implemented second part of the internal-instance logic: the existing configuration for the internal instance provides kubectl access by default. Therefore, if the scenario instructs not having it, the config enabling access to the cluster should be deleted.

Note that I haven't enabled the use of yq as the apt repository is failing and I don't want to introduce unnecessary issues.

Also, on the implemented logic for the internal-instance, what are the thoughts on having this being executed regardless of the starting_point function call given that the user can ssh into the internal instance without it?

raoulmillais commented on Dockerfile:
```diff
@@ -51,23 +51,25 @@ ARG lint_user=lint
+
 # Setup non-root lint user
-    && useradd -ms /bin/bash ${lint_user} \
+RUN useradd -ms /bin/bash ${lint_user} \
```
Is there a reason you split this into 2 RUN commands?  Each RUN adds a new layer when building in creasing disk space usage and slowing down the build a little.

s-irvine replied:
Merged back into one command

raoulmillais commented:
I've approved the previously submitted change requests and added on  new little query.  I will  tag and push the branch now so we can test

raoulmillais commented:
Another minor improvement:: task 1 has no starting-point but task 2 does.. the user is on task 1.  Should we elaborate in the message on how the user should proceed (mentioning the ssh helpers) instead of just saying "There's no starting point"?

abdullahgarcia commented on attack/scripts/bashrc:
```diff
@@ -65,18 +65,47 @@ readonly -f info
 
     if [[ "$KUBECTL_ACCESS" == "false" ]]; then
-      ssh "$INTERNAL_NODE_IP" rm -r ~/.kube
+      ssh "$INTERNAL_NODE_IP" mv -r ~/.kube /var/local/.kube
```
Is this command/flag valid: "mv -r" ?

s-irvine replied:
Nope, will fix.

abdullahgarcia commented on attack/scripts/bashrc:
```diff
@@ -65,18 +65,47 @@ readonly -f info
-      ssh "$INTERNAL_NODE_IP" rm -r ~/.kube
+      ssh "$INTERNAL_NODE_IP" mv -r ~/.kube /var/local/.kube
+    elif [[ ! -d ~/.kube ]]; then
+      ssh "$INTERNAL_NODE_IP" mv -r /var/local/.kube ~/.kube
```
Is this command/flag valid: "mv -r" ?

raoulmillais commented:
This LGTM from reading the code. I'll push the tag to manually verify the behaviour

s-irvine commented:
Fixes #79 

Creating #180
